유태식 - 자동차 게임

taesikyoo opened the pull request:
아쉬운 부분들은
- 구현하지 못한 기능들이 있습니다.(자동차 이름이 다섯글자 이하, 자동차의  이동 시각화 등)
- commit log 요구사항대로 기능단위로 commit하지 못했습니다.(중간 중간 커밋할 때를 놓쳤습니다.)
- 기능 목록에 적은 것 외에 리팩토링하고 추상화할 요소들이 많이 보이는데 다 하지 못했습니다.
- 테스트 코드를 어떻게 짜야할지 몰라서 많이 헤맸습니다.

그래도 배운 점은
- 테스트를 구현한 후 실제 코드를 작성하는 흐름을 조금 익혔습니다.
- 광일쌤하고 덕수쌤이 라이브코딩하실 때 왜 그렇게 짰는지 더 이해가 잘 가게 되었습니다.
- 일단 그래도 프로그램이 돌아가니까 넘 기쁩니다,,,

kwangilcho-at-woowabros committed:
Add assertj dependency and sample test

- Add assertj
- Add sample test to guide using assertj

kwangilcho committed:
Add assertj dependency and sample test (#1)

- Add assertj
- Add sample test to guide using assertj

Co-authored-by: 조광일 <kwangil.cho@woowahan.com>

chapitak committed:
Add fist week works

kwangilcho committed:
Merge pull request #2 from chapitak/김정교

첫 주차 세션에서  작업 내용입니다

taesikyoo committed:
upstream 연습

taesikyoo committed:
input 받기

taesikyoo committed:
자동차 이름과 시도 횟수를 한 메서드에서 입력받기
CarGameInfo 클래스 만들기

taesikyoo committed:
Merge branch '유태식' of https://github.com/kiworkshop/java-racingcar into 유태식

taesikyoo committed:
Car 클래스 만들기

taesikyoo committed:
CarGame과 CarGameInfo 합치기
Main에 getGameInfo 메서드 만들기

taesikyoo committed:
자동차 이름 입력받기(이름 여러개 입력받기)

taesikyoo committed:
실행 결과 하나 출력하기

taesikyoo committed:
랜덤 숫자 만들기
Car.forward() 구현하기

taesikyoo committed:
최종 우승자 출력하기

wbluke commented on src/main/java/CarGame.java:
```diff
@@ -1,65 +1,77 @@
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.List;
 import java.util.Scanner;
```
모든 파일에서 전체적으로 사용하지 않는 package들이 많이 보이는데요, 맥 기준으로 `option + cmd + L`(윈도우 : `ctrl + alt + L`) 하시면 import 정리를 전부 해줍니다. 왼쪽에 패키지명에 커서를 두고 해당 단축키를 사용하시면 해당 패키지 하위에 있는 package 정리 기능을 사용하실 수 있습니다. (코드 컨벤션도 맞춰줍니다.)

개인적으로 코드 컨벤션과 사용하지 않는 패키지 정리는 클린코드의 첫걸음이라고 생각합니다ㅎㅎ 저는 커밋 찍기 전에 매번 해당 단축키로 파일들을 정리하는 습관이 있는데요, 이번 스터디 하시면서 좋은 습관으로 가져가시면 좋을 것 같아요! :)

wbluke commented on src/main/java/CarGame.java:
```diff
@@ -1,65 +1,77 @@
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.List;
 import java.util.Scanner;
```
전체적으로 사용하지 않는 패키지들이 많이 보이는데요. 맥 기준으로 `option + cmd + L`(윈도우 : `ctrl + alt + L`)하시면 코드컨벤션도 맞춰주고 사용하지 않는 패키지들도 정리해 줍니다.
왼쪽 패키지명에 커서를 두시고 해당 단축키를 사용하시면 하위 패키지들에 대해 한방에 전부 정리해 줍니다.

개인적으로 코드 컨벤션을 맞추는 것과 사용하지 않는 패키지들을 정리하는 것은 클린 코드의 첫걸음이라고 생각합니다.ㅎㅎ 저는 커밋을 찍기 전에 늘 해당 단축키로 코드를 한번 정리하는 시간을 가지는데요, 이번 스터디를 통해 좋은 습관을 가져가시면 좋을 것 같아요 :)

taesikyoo replied:
메일이 갈까요?

wbluke commented on src/main/java/main.java:
```diff
@@ -1,8 +1,22 @@
+    public static CarGame getGameInfo() {
+        Scanner sc = new Scanner(System.in);
+        System.out.println("경주할 자동차 이름을 입력하세요.(이름은 쉼표(,)기준으로 구분)");
+        List<String> carNames = Arrays.asList(sc.nextLine().split(","));
```
```suggestion
        String inputCarName = sc.nextLine();
        String[] tokens = inputCarName.split(",");
        List<String> carNames = Arrays.asList(tokens);
```

한줄에 너무 많은 작업을 하는 것 같아요. `디미터 법칙`이라는 키워드로 한번 찾아보세요 :)

wbluke commented on src/main/java/main.java:
```diff
@@ -1,8 +1,22 @@
-        CarGame game = new CarGame();
-        game.start();
+
+    public static CarGame getGameInfo() {
```
하나의 메소드에서 세부 private 메소드들로 쪼갤 수 있는 흐름이 보이는 것 같아요ㅎㅎ 
기능을 쪼개고 이름을 붙여서 추상화를 해보는 것은 어떨까요? :)

wbluke commented on src/test/java/TestCar.java:
```diff
@@ -0,0 +1,31 @@
+    @Test
+    public void testEquality() {
+        Car car1 = new Car();
+        car1.setName("hyundai");
```
setter 사용은 최대한 지양하는 것이 좋습니다. 조금 강하게 표현하면 아예 금지된 방식이라 여기셔도 좋습니다. 
(이미 스프링을 포함한 많은 프레임워크에서는 setter를 보실 수 있지만, 적어도 내가 짠 코드에서는 setter가 없도록 하는 것이 좋습니다.)
왜 setter를 사용하면 안될까요? 한번 고민해보시면 좋을 것 같아요. :)

지금 test에서는 기본 생성자로 Car를 생성하고, 이름을 후에 세팅해주는 식으로 객체를 만들고 있는데요.
Car에 기본 생성자를 없애고, 이름으로 생성하는 생성자만 있어도 괜찮지 않을까요? :)

wbluke commented on src/main/java/Car.java:
```diff
@@ -1,4 +1,49 @@
+
+    public void forward() {
+        progress++;
+    }
```
👍

wbluke commented on README.md:
```diff
@@ -1 +1,26 @@
+    - ~~CarGameResult 클래스 만들기~~
+    
+- ~~최종 우승자 출력하기~~
+    - progress를 시각화하기 
```
👍

wbluke commented on src/main/java/CarGame.java:
```diff
@@ -1,65 +1,77 @@
-        Scanner sc = new Scanner(System.in);
-        System.out.println("경주할 자동차 이름을 입력하세요.(이름은 쉼표(,)기준으로 구분).");
-        this.inputString =  sc.nextLine();
+    public static int generateRandomNumber() {
```
해당 함수는 private으로 둬도 될 것 같아요. 그리고 메소드의 순서는 public 메소드가 나온 후에, 해당 메소드 내부의 private 메소드가 차례로 나와야 합니다. 그래야 자연스럽게 읽히겠죠? ㅎㅎ
이 메소드는 아래에 있는 run() 메소드에서 사용하기 때문에 run() 메소드 아래에 위치하면 좋을 것 같아요. 
다른 메소드들도 한번 체크해보시면 좋겠네요 :)

wbluke commented on src/main/java/CarGame.java:
```diff
@@ -1,65 +1,77 @@
+    public CarGameResult run() {
+        List<Car> newCars = new ArrayList<>();
+        for (Car car : cars) {
+            if (CarGame.generateRandomNumber() >= 4) {car.forward();}
```
4 라는 숫자는 매직넘버네요 :) 상수로 관리하는 것이 어떨까요?

wbluke commented on src/main/java/CarGame.java:
```diff
@@ -1,65 +1,77 @@
+            newCars.add(car);
+        }
+        return new CarGameResult(newCars);
     }
```
run() 메소드가 두 가지 일을 하고 있는 것으로 보여요. 
1. 자동차가 진행할 것인지 판단하는 일
2. 실제로 자동차를 진행시키는 일

심지어 랜덤생성 로직이 끼어 있기 때문에 해당 메소드가 테스트 불가능한 구조인 것 같네요.ㅎㅎ
테스트 불가능한 로직을 외부로 분리하고, 해당 메소드를 테스트 가능하도록 만들 수 있을까요? :)

wbluke commented on src/test/java/TestCarGame.java:
```diff
@@ -0,0 +1,74 @@
+        System.out.println(CarGame.generateRandomNumber());
+        System.out.println(CarGame.generateRandomNumber());
+        System.out.println(CarGame.generateRandomNumber());
+    }
```
assert문을 쓰지 않고 console에 단순하게 찍어보는 테스트는 올바른 테스트 로직이 아닙니다 :)
위에 있는 테스트들도 마찬가지네요 ㅎㅎ

wbluke commented on src/main/java/CarGame.java:
```diff
@@ -1,65 +1,77 @@
+            }
+        }
+        System.out.println(String.join(", ", winners) + "가 최종 우승했습니다.");
     }
```
콘솔에 결과를 출력하는 로직은 CarGame 클래스에 있기에는 해당 클래스의 역할이 너무 많은 것 같아요.
OutputView 등으로 클래스를 나누어서 책임을 분리해 보면 어떨까요?
추가로 Scanner로 입력받는 역할도 InputView 등으로 나눠볼 수 있을 것 같아요 :)

wbluke reviewed (changes requested):
안녕하세요 태식 선생님!
테스트 구현 경험이 없으실텐데 고생 많으셨습니다ㅎㅎ 고생의 흔적들이 많이 보이네요.
앞으로 스터디 통해서 계속 꾸준히 연습하면 더 좋은 코드를 짜실  수 있을 겁니다.ㅎㅎ
몇 가지 코멘트 남겼는데요, 한번 보시고 모르겠는 부분은 편하게 댓글 남기거나 DM 주세요! :)

taesikyoo committed:
docs: 구현할 기능 목록 작성

- README.md에 구현 기능 목록 작성

taesikyoo committed:
refactor: main의 일부 코드를 InputView로 추출

- 디렉토리 구조를 MVC 패턴으로 수정
- main에서 console로 input을 받는 코드들을 View의 InputView 클래스에
이주시킴
- main의 이름은 RacingCarGame으로 변경, car 역시 RacingCar로 이름 수정

taesikyoo committed:
feat: RacingCars 객체 추가

- List에 RacingCar를 담고 있는 일급 객체 RacingCars를 추가

taesikyoo committed:
feat: RacingCarService에 RacingCars를 만드는 로직 추가

- delimeter를 기준으로 자동차 이름을 분리하여 RacingCar의 List를 생성
- 이 List를 RacingCars의 생성자에 전달

taesikyoo committed:
test: RacingCarsTest 추가

- 중복된 car가 있는 경우 illegalArgumentException을 throw하도록
RacingCars를 수정

taesikyoo committed:
feat: RacingCar의 이름 길이를 검사하는 기능 추가

- RacingCar에서 nameIsLongerThanMaxLength()를 통해 이름 길이를 검사

taesikyoo committed:
feat: ProceedingStrategy 인터페이스 추가

- 테스트가 가능한 경우와 불가능한 경우를 분리시킴

taesikyoo committed:
refactor: RacingCars와 RacingCar의 moveForward 메서드 수정

- ProceedingStrategy를 매개변수로 추가

taesikyoo committed:
test: RacingCarTest에 moveForward() 테스트를 추가

- ManualProceedingStrategy를 이용

taesikyoo committed:
feat: RacingCarSnapShot 객체 추가 및 RacingCarGameHistory 수정

- RacingCarGameHistory에 takeSnapShots()와 takeSnapShot() 추가
- takeSnapShot()에서 RacingCarSnapShot 객체를 생성하고 이 객체는 인자로
받은 car의 인스턴스 필드를 deep copy함

taesikyoo committed:
feat: RacingCarGameResult에 record 메서드 추가

- record()는 인자로 받은 history를 인스턴스 필드의 histories에 저장

taesikyoo committed:
test: RacingCarGameSnapShotTest 추가

- 스냅샷이 RacingCar를 잘 복사하는지와 불변하는지 테스트

taesikyoo committed:
test: RacingCarGameHistoryTest 추가

- History객체 안의 스냅샷 수가 실제 자동차 수와 동일한지 테스트

taesikyoo committed:
feat: OutputView에서 RacingGameResult를 출력하는 기능 추가

taesikyoo committed:
feat: RacingCarGameResult와 History에서 우승자를 찾는 기능 추가

- OutputView에서 Result의 findWinnersIn(lastRound)를 호출
- Result에서 History의 findWinners()를 호출

woowabean commented on src/main/java/racingcar/domain/car/RacingCars.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.*;
+import java.util.stream.Stream;
+
+public class RacingCars {
```
오우 일급컬렉션을 적용하셨네요! 
다음 미션이나 다다음 미션쯤 일급컬렉션 피드백을 드릴까 했었는데, 저번 모임때 들으시고 바로 적용하시다니 아주 멋집니다 👍👍👍

woowabean commented on src/main/java/racingcar/domain/car/RacingCars.java:
```diff
@@ -0,0 +1,50 @@
+
+public class RacingCars {
+
+    List<RacingCar> cars = new ArrayList<>();
```
요 초기화는 지금 인텔리제이에서 회색으로 표기가 되어있을텐데, 생성자에서 초기화를 해주기 때문에 다른 의도가 없으시다면 없애셔도 될거같아요ㅎㅎ

woowabean commented on src/main/java/racingcar/domain/car/RacingCars.java:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+
+    public Stream<RacingCar> stream() {
+        return cars.stream();
+    }
```
일급컬렉션 내의 데이터를 가공한 값을 외부에서 사용하기 위해 해당 로직이 나온 것으로 보이는데요,
가공의 여지를 주는 stream을 외부로 던지는 것 보다, 일급컬렉션 내부에서 가공을 끝내고 결과값만 외부로 보내는 것이 좋습니다.
(객체지향의 개념에 대해서 생각해보면 이해가 가실거에요)

https://jojoldu.tistory.com/412
일급컬렉션에 대해서는 앞으로도 고민할 기회가 많을거에요. 해당 글을 레퍼런스 삼아 다음, 다다음 미션에서도 적용해보시면 좋을 것 같습니다~!

woowabean commented on src/main/java/racingcar/domain/RacingCarSnapShot.java:
```diff
@@ -0,0 +1,22 @@
+    private final String name;
+    private final int position;
+
+    public RacingCarSnapShot(final RacingCar car) {
+        this.name = car.getName();
+        this.position = car.getPosition();
+    }
```
스냅샷 👍 

RacingCar의 상태를 보존하기 위해 해당 자동차의 데이터를 전부 꺼내서 저장하는 형식인데요,
만약 차의 필드가 정말정말 많아지면 어떻게 될까요? (이런 극단적인 사고가 설계할 때 도움이 많이 됩니다)

자동차를 그대로 카피(깊은 복사) 떠서 저장하는 방법도 있을 것 같아요ㅎㅎ

woowabean commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,63 @@
+package racingcar.view;
+
+import com.sun.deploy.util.StringUtils;
+import racingcar.domain.RacingCarGameHistory;
+import racingcar.domain.RacingCarGameResult;
+import racingcar.domain.RacingCarSnapShot;
+import racingcar.domain.car.RacingCar;
```
사용하지 않는 import 제거~

woowabean commented on src/test/java/racingcar/domain/car/RacingCarTest.java:
```diff
@@ -0,0 +1,42 @@
+        RacingCar car2 = new RacingCar("abcde");
+
+        // when
+        Throwable thrown = catchThrowable(() -> {
+            new RacingCar("abcdef");
+        });
+
+        // then
+        assertEquals("abcd", car1.getName());
+        assertEquals("abcde", car2.getName());
+        assertThat(thrown).isInstanceOf(IllegalArgumentException.class);
```
```java
assertThrows(IllegalArgumentException.class, () -> new RacingCar("abcdef"));
```
예외 테스트는 이렇게 한번에도 할 수 있습니다 :)

woowabean commented on src/main/java/racingcar/domain/proceedingstrategy/ProceedingStrategy.java:
```diff
@@ -0,0 +1,6 @@
+
+public interface ProceedingStrategy {
+
+    public boolean isProceedable();
```
인터페이스에는 public 접근제어자가 무의미합니다 :)

woowabean commented on src/main/java/racingcar/domain/proceedingstrategy/ManualProceedingStrategy.java:
```diff
@@ -0,0 +1,9 @@
+package racingcar.domain.proceedingstrategy;
+
+public class ManualProceedingStrategy implements ProceedingStrategy{
```
프로덕션 코드가 아닌 테스트에서만 쓰는 클래스는 테스트쪽에 두시거나, 테스트 클래스 내에서 생성해서 쓰는 것이 맞습니다ㅎㅎ
그리고 해당 전략은 굳이 클래스를 만들지 않아도 람다로 해결이 가능합니다 :)
```java
() -> true;
```

woowabean reviewed (approved):
태식 선생님~ 리뷰를 아주 잘 반영해주셨네요 💯 
온오프라인 리뷰 모두 공들여 반영해주신 것 같아 감사합니다.
일급컬렉션이나 테스트 관련해서 몇 가지 피드백 남겨드리는데요, 이번 미션은 많이 반영하셨으니 이만 approve 하고
다음 미션에서 피드백 참고하셔서 반영해보시면 좋을 것 같습니다. 수고 많으셨습니다~~!!

wbluke reviewed (approved):
잘못해서 회사계정으로 리뷰 남겼네요ㅠㅠ ㅋㅋ 원래 계정으로도 approve 하겠습니다
